File: tools/chess-player/src/main/java/com/cooperate/chessplayer/service/PgnExporter.java
package com.cooperate.chessplayer.service;

import com.cooperate.chessplayer.config.ChessPlayerProperties;
import com.cooperate.chessplayer.model.GameOutcome;
import com.cooperate.chessplayer.model.MatchState;
import com.github.bhlangonijr.chesslib.Side;
import com.github.bhlangonijr.chesslib.move.Move;
import com.github.bhlangonijr.chesslib.move.MoveList;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;

import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardOpenOption;

@Component
public class PgnExporter {
  private static final Logger log = LoggerFactory.getLogger(PgnExporter.class);

  private final ChessPlayerProperties properties;

  public PgnExporter(ChessPlayerProperties properties) {
    this.properties = properties;
  }

  public void export(MatchState state) {
    if (state == null || !properties.isPgnExportEnabled() || state.getMatchId() == null) {
      return;
    }

    Path dir = Path.of(properties.getPgnExportDir());
    Path pgnPath = dir.resolve(state.getMatchId() + ".pgn");
    try {
      Files.createDirectories(dir);
      String pgn = buildPgn(state);
      Files.writeString(
          pgnPath,
          pgn,
          StandardOpenOption.CREATE,
          StandardOpenOption.TRUNCATE_EXISTING,
          StandardOpenOption.WRITE
      );
      log.info("Exported PGN for match {} to {}", state.getMatchId(), pgnPath);
    } catch (Exception e) {
      log.warn("Failed to export PGN for match {}", state.getMatchId(), e);
    }
  }

  private String buildPgn(MatchState state) {
    MoveList moveList = new MoveList();
    Side side = Side.WHITE;
    for (String uci : state.getMoveHistoryUci()) {
      moveList.add(new Move(uci, side));
      side = side == Side.WHITE ? Side.BLACK : Side.WHITE;
    }

    String moveText;
    try {
      moveText = moveList.toSanWithMoveNumbers();
    } catch (Exception e) {
      moveText = String.join(" ", state.getMoveHistoryUci());
    }

    String result = switch (state.getOutcome()) {
      case WHITE_WIN -> "1-0";
      case BLACK_WIN -> "0-1";
      case DRAW -> "1/2-1/2";
      case ONGOING -> "*";
    };

    StringBuilder sb = new StringBuilder();
    sb.append("[Event \"ACP Chess AI Match\"]\n");
    sb.append("[Site \"ACP Direct\"]\n");
    sb.append("[Round \"-\"]\n");
    sb.append("[White \"AI-White\"]\n");
    sb.append("[Black \"AI-Black\"]\n");
    sb.append("[Result \"").append(result).append("\"]\n");
    sb.append("[MatchId \"").append(state.getMatchId()).append("\"]\n");
    sb.append("[SessionId \"").append(state.getUcwId()).append("\"]\n");
    sb.append("\n");
    sb.append(moveText);
    if (!moveText.isBlank()) {
      sb.append(' ');
    }
    sb.append(result).append('\n');
    return sb.toString();
  }
}

File: tools/python-chess-player/app/pgn_exporter.py
from __future__ import annotations

import logging
from pathlib import Path

import chess
import chess.pgn

from .config import AppConfig
from .models import GameOutcome, MatchState


LOG = logging.getLogger(__name__)


class PgnExporter:
    def __init__(self, config: AppConfig) -> None:
        self._config = config

    def export(self, state: MatchState | None) -> None:
        if (
            state is None
            or state.matchId is None
            or not self._config.pgn_export_enabled
        ):
            return

        output_dir = Path(self._config.pgn_export_dir)
        output_path = output_dir / f"{state.matchId}.pgn"
        try:
            output_dir.mkdir(parents=True, exist_ok=True)
            output_path.write_text(self._build_pgn(state), encoding="utf-8")
            LOG.info("Exported PGN for match %s to %s", state.matchId, output_path)
        except Exception:
            LOG.warning("Failed to export PGN for match %s", state.matchId, exc_info=True)

    @staticmethod
    def _build_pgn(state: MatchState) -> str:
        board = chess.Board()
        game = chess.pgn.Game()
        game.headers["Event"] = "ACP Chess AI Match"
        game.headers["Site"] = "ACP Direct"
        game.headers["Round"] = "-"
        game.headers["White"] = "AI-White"
        game.headers["Black"] = "AI-Black"
        game.headers["Result"] = PgnExporter._result_token(state.outcome)
        game.headers["MatchId"] = str(state.matchId)
        game.headers["SessionId"] = str(state.ucwId)

        node = game
        for uci in state.moveHistoryUci or []:
            if not uci or len(uci) < 4:
                break
            try:
                move = chess.Move.from_uci(uci)
            except ValueError:
                break
            if move not in board.legal_moves:
                break
            node = node.add_variation(move)
            board.push(move)

        exporter = chess.pgn.StringExporter(headers=True, variations=False, comments=False)
        return game.accept(exporter) + "\n"

    @staticmethod
    def _result_token(outcome: GameOutcome | None) -> str:
        if outcome is GameOutcome.WHITE_WIN:
            return "1-0"
        if outcome is GameOutcome.BLACK_WIN:
            return "0-1"
        if outcome is GameOutcome.DRAW:
            return "1/2-1/2"
        return "*"
